core/src/main/java/io/grpc/Metadata.java

```diff
@@ -249,18 +337,24 @@ public boolean containsKey(Key<?> key) {
   public <T> boolean remove(Key<T> key, T value) {
     Preconditions.checkNotNull(key, "key");
     Preconditions.checkNotNull(value, "value");
-    List<MetadataEntry> values = store.get(key.name());
-    if (values == null) {
-      return false;
-    }
-    for (int i = 0; i < values.size(); i++) {
-      MetadataEntry entry = values.get(i);
-      if (!value.equals(entry.getParsed(key))) {
-        continue;
+    for (int i = 0; i < size; i++) {
+      if (Arrays.equals(key.asciiName(), name(i))) {
```

Comment:
nit: might save a level of indentation by using negation. also in the other methods.